File: 02-java-exercises/4-palindrome-java/src/Palindrome.java
public class Palindrome {

    public static void main(String[] args) {

        String wordToSearch = "abnonba";

        System.out.println(wordToSearch + " is a palindrome: " + isPalindromeIterative(wordToSearch) + ".");
        System.out.println(wordToSearch + " is a palindrome: " + isPalindromeRecursive(wordToSearch) + ".");
    }

    /* PALINDROME (iterative) */
    public static boolean isPalindromeIterative(String word) {
        int finalChar = word.length() - 1;

        if (word.length() == 1) {
            return true;
        }
        for (int i = 0; i < Math.floor(word.length() / 2); i++) {
            if (word.charAt(i) != word.charAt(finalChar - i)) {
                return false;
            }
        }
        return true;
    }


    /* PALINDROME (recursive) */
    public static boolean isPalindromeRecursive(String word) {
        int beg = 0;
        int end = word.length() - 1;

        if (word.length() <= 1) {
            return true;
        }
        if (word.charAt(beg) != word.charAt(end)) {
            return false;
        }

        String subs = word.substring(beg + 1, end);
        return isPalindromeRecursive(subs);
    }
}

File: 4-palindrome/palindrome.js
/* PALINDROME */

var word = "annnnnnna";

console.log("// Palindrome (iterative) //");

function isPalindromeIterative(wordToCheck) {

  var finalChar = wordToCheck.length - 1;

  if (wordToCheck.length === 1) {
    return true;
  }

  for (var i=0; i<Math.floor(wordToCheck.length / 2); i++) {
    if (wordToCheck[i] !== wordToCheck[finalChar-i] ) {
      return false;

    }
  } return true;

}

console.log(word + " is a palindrome: " + isPalindromeIterative(word) +".");


console.log("\n");


console.log("// Palindrome (recursive) //");

function isPalindromeRecursive(wordToCheck) {

  var beg = 0;
  var end = wordToCheck.length - 1;

  if (wordToCheck.length <= 1) {
    return true;
  }

  if (wordToCheck[beg] !== wordToCheck[end]) {
    return false;
  }

  var subs = wordToCheck.substring(beg + 1, end);
  return isPalindromeRecursive(subs);
}

console.log(word + " is a palindrome: " + isPalindromeRecursive(word) +".");
